rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,552 @@
+
+Kubernetes Pods are, by definition, expected to be stateless. This means that each time a Pod is recycled because it was restarted, deleted, upgraded, or moved to another node, the state that was written to its filesystem is lost.
+
+One way to overcome this problem is to use Persistent Volumes. PV is a Kubernetes feature that mounts a storage volume in the container filesystem, whose lifecycle is independent of the Pod that mounts it. Kubernetes’ PV storage has its own drawbacks. Persistent Volumes are very distinct between different cloud vendors. Some providers lock volumes to be mounted in the same zone they were created, meaning that if Teleport agent Pod is recycled, it must be maintained in the same zone it was created. This creates operational issues and for on-premises deployments might be difficult to manage.
+
+Another possibility is that the agent might use the invite token each time the pod is recycled, meaning that it issues a join request to the cluster every time it starts and receives a new identity from the Auth service. This means that the invite token must be a static/long-lived token, otherwise after a while the agent could not register himself in the cluster because the invite token expired. This approach is not recommended and might cause security flaws because the join token can be stolen or guessed.
+
+One solution that might address all the issues referenced above is to allow the agent to use Kubernetes secrets as storage backend for process state. This allows not only that the agent is able to run stateless depending only on native objects generally available in any Kubernetes cluster, but also that the agent might support dynamic short-lived invite tokens with no dependency on external storage.
+
+## Details
+
+### Secret creation and lifecycle
+
+The agent creates, reads and updates the Kubernetes Secrets.
+
+The agent will never issue a delete request to destroy the secret. Helm’s `post-delete` hook achieves this once the operator executes `helm uninstall`.
+
+#### Secret content
+
+Once agent updates the secret, it will have the following structure:
+
+```yaml
+apiVersion: v1
+data: |
+        {
+            # {role} might be kube, app,... if multiple roles are defined for the agent then, 
+            multiple entries are created each one holding its identity
+            "/ids/{role}/current": {
+                "kind": "identity",
+                "version": "v2",
+                "metadata": {
+                    "name": "current"
+                },
+                "spec": {
+                    "key": "{key_content}",
+                    "ssh_cert": "{ssh_cert_content}",
+                    "tls_cert": "{tls_cert_content}",
+                    "tls_ca_certs": ["{tls_ca_certs}"],
+                    "ssh_ca_certs": ["{ssh_ca_certs}"]
+                }
```

Comment:
Hmm, I'm now thinking that this secret format may make it difficult to teach Machine ID to periodically refresh agents' host certs. I think the idea was to eventually teach Machine ID to regenerate secrets by a particular label selector containing certificates.

@timothyb89 Have you thought about this previously? The context is that we're moving the agent's identity into Kubernetes secrets to properly support dynamic join tokens, and we'd probably rather come up with a way to store it in Kubernetes which would be compatible with Machine ID.

Reply:
@timothyb89 , is MachineID expected to write content into the disk if the secret is mounted locally? 

If yes, unfortunately, it is not possible to write into mounted configmaps or secrets in Kubernetes (they are read-only objects) and any update must happen via API call

Reply:
Right, the thinking was that Machine ID would be running as it's own deployment and writing to a secret which other pods could then consume as desired (with k8s RBAC used to control which other workloads could access the secret). We designed Machine ID to have multiple read/write backends so we'd implement one to read/write secrets via the API rather than the filesystem backend it uses today.

All that's said with the huge caveat of, this is just what's been rattling around in my head as a long term plan, we haven't really planned anything here formally.

Reply:
Yes, so basically the idea for the evolution of this feature was to eventually be able to deploy Machine ID in a Kubernetes cluster and be able to tell it

```
periodically rotate the secrets in namespace "teleport" matching label a=b containing Teleport agent identity
```

and potentially extend it to support issuing individual certs signed by Teleport CA and saving them to secrets matching labels also.

With this in mind, what do you think about splitting this secret up into 3:

- Secret containing current identity, in the same format as it already is, the json basically.
- Secret containing replacement identity, will only be present during CA rotation.
- Secret containing rotation state just because we have to keep it in K8s also with this approach.

Then in future we'll be able to add support to Machine ID to rotate "agent identity" secret and it will plug in nicely into this model. I don't think assembling the identity in Teleport process would be much more complicated with 3 secrets vs 1.

@tigrato @timothyb89 What do you guys think?

Reply:
@r0mant, why do you think the rotation is different with 1 vs 3 secrets? 

As far as I know, the structure is the same as in the SQLite database and each entry is controlled by the key defined in the `ProcessStorage` depending on the action (`WriteStatus`, `WriteIdentity`, `ReadIdentity` ...) that is passed into the storage backend.

If Machine ID can plug different backend storages and is able to use the Kubernetes backend storage, it's exactly the same as opening and operating the local SQLite DB.  The only difference is that when `NewKubeSecretStatusStorage` is called instead of providing the local path you must provide the `TELEPORT_REPLICA_NAME`.

The issue here is probably how would we notify the agent that  Machine ID decided to rotate its credentials.

Reply:
I just think it would be "cleaner" if the identity was stored as a separate secret. Then it would be easy to reason about e.g. "Machine ID can rotate secrets of types 'teleport identity', 'host certificate', etc.". And the refresh process would be the same for each secret "type" - create new secret, remove old secret. Versus if we keep the current structure, Machine ID will have to know how to keep both replacement and old identity in 1 secret. Definitely doable, but just seems like the other way is just cleaner design.

> The issue here is probably how would we notify the agent that Machine ID decided to rotate its credentials.

Machine ID can just be configured to restart the pod that uses the certificate automatically. There are already controllers in Kubernetes that do that. But that's probably a question for later when we actually get to Machine ID integration. For now let's just make sure the the design allows for it.

Reply:
When `ProcessStorage` executes, for instance, the `WriteIdentity` function it calls the `Write` method in storage and passes the key, `backend.Key` with the format of `/ids/{role}/current` or `/ids/{role}/replacement`. 

With your change, it means that the Kubernetes storage has to check the Key type, i.e.  `current` or `replacement`, `status`  and map the data into the correct secret. This is a bit weird and might require changes if, in the future, Teleport stores other event types in dbs or if the keys structure changes.

I am unsure how Machine ID works under the hood, but I assume it uses a `backend.Backend` interface implementation to interact with storage. If that's the case it already receives the `backend.Keys` in the format I showed before and it's just a matter o plug and unplugging different backend storage implementations.

@timothyb89 can you give some intel on how would you prefer it?

Reply:
I guess I'm ultimately ambivalent here given our Kubernetes integration doesn't exist yet and we'll have a lot of freedom to implement whatever custom logic is needed.

My two concerns are:
 * Each key inside the secret contains nested fields which are ultimately what we want to update (specifically the key data / certs / etc), meaning we'd need to deserialize, update, and finally reserialize values. Certainly not impossible but does add a bit of complexity.
 * The secret contains some internal state so I expect there to be multiple writers. It would be okay if it was just Machine ID writing the secret so we could just pave over everything with the latest certs, but with multiple writers we'll need to update the secret as atomically as possible on all writers' ends. I think there's some options around server side apply / resource versions / merge strategies we can use to make this work. Maybe there's already some work done on this?

We'll almost certainly be using a different secret format for users who just want to use Machine ID in k8s but it should be simple enough to have a separate backend impl specifically for this use-case.

> The issue here is probably how would we notify the agent that Machine ID decided to rotate its credentials.

There are a few options beyond just restarting the agent with an API call: if the secret is volume mounted we can get update events using `inotify` or just plain filesystem polling, or use a k8s API watch.

Reply:
> I guess I'm ultimately ambivalent here given our Kubernetes integration doesn't exist yet and we'll have a lot of freedom to implement whatever custom logic is needed.
> 
> My two concerns are:
> 
> * Each key inside the secret contains nested fields which are ultimately what we want to update (specifically the key data / certs / etc), meaning we'd need to deserialize, update, and finally reserialize values. Certainly not impossible but does add a bit of complexity.
> * The secret contains some internal state so I expect there to be multiple writers. It would be okay if it was just Machine ID writing the secret so we could just pave over everything with the latest certs, but with multiple writers we'll need to update the secret as atomically as possible on all writers' ends. I think there's some options around server side apply / resource versions / merge strategies we can use to make this work. Maybe there's already some work done on this?
> 
> We'll almost certainly be using a different secret format for users who just want to use Machine ID in k8s but it should be simple enough to have a separate backend impl specifically for this use-case.
> 
> > The issue here is probably how would we notify the agent that Machine ID decided to rotate its credentials.
> 
> There are a few options beyond just restarting the agent with an API call: if the secret is volume mounted we can get update events using `inotify` or just plain filesystem polling, or use a k8s API watch.

Regarding the two concerns:

- The nested fields have exactly the same structure as the JSON we store in  `SQLite`. 
Yes, it will require something like Reading + Unmarshling + Update Fields + Marshaling + Update Secret if MachineID just updates a single value inside the nested JSON but most of the actions are already hidden by the backend storage.
If MachineID wants to replace the JSON for a key, you need to call `Update(key, data)` and the backend will read the secret data, unmarshal it,  merge the updated key data, and update the secret. 
The exact same behavior we have with `SQLite` for Identities. 

- The KubeSecret backend will already use the versioning control. Kubernetes has native support for it, so it will be quite simple to make sure multiple agents/bots do not update the same asset concurrently.

Reply:
In that case I think we can work with the existing design!

Reply:
Sorry guys to keep bike-shedding on this, but I keep looking at this format and can't bring myself to liking it. Especially the part with using keys that for some reason try to emulate the backend keys inside the secret's json. I don't understand why this should in any way depend on which key paths are being used in the backend. I understand how _maybe_ it would make implementation a little easier but I'm not sure we should optimize for that.

I think we should reconsider splitting this up in 3 secrets.

Reply:
@r0mant,

The keys are not trying to emulate the backend. 

- Keys are managed by `ProcessStorage` methods:

https://github.com/gravitational/teleport/blob/ef9937a83ed633bf8831197572d2b5639ef25f61/lib/auth/state.go#L164-L170

`WriteIdentity`, `ReadIdentity`, `GetState`, `CreateState` and `WriteState` methods generate the keys before calling the storage based on constants defined in the file.

- Storage already receives the key built when it's called and does not require any mapping between `backend.Key` generated by `ProcessStorage` and the secret that stores that information.

- Secret storage can work as pluggable storage for identities/status and can be clueless about how `ProcessStorage` generates its keys as long as they can be converted to valid strings. This means that when we create the `ProcessStorage` we can switch the backend like this:
```go
// NewProcessStorage returns a new instance of the process storage.
func NewProcessStorage(ctx context.Context, path string) (*ProcessStorage, error) {
	var (
		storage IdentityBackend
		err     error
	)
	if path == "" {
		return nil, trace.BadParameter("missing parameter path")
	}

	if kubernetes.InKubeCluster() {
		storage, err = kubernetes.New(ctx)
                 // compatibility layer if secret does not exist but local storage exists!
	} else {
		storage, err = lite.NewWithConfig(ctx, lite.Config{
			Path:      path,
			EventsOff: true,
			Sync:      lite.SyncFull,
		})
	}
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return &ProcessStorage{IdentityBackend: storage}, nil
}
```

- `MachineID` already builds the `backend.Key` in order to load the identity and store them somewhere.

Besides, check this use case and let me know your thoughts.

Let's assume a configuration where the agent is running with roles = [kube, app, db] because it proxies Grafana and some DB inside the cluster.

With this config, and since each role has its own identity, status, and replacement keys, it means that if we are using SQLite backend we end up with the following keys in the `kv` table, assuming that a rotation happened:

- `/ids/app/current`
- `/ids/kube/current`
- `/ids/db/current`
- `/ids/app/replacement`
- `/ids/kube/replacement`
- `/ids/db/replacement`
- `/states/app/state`
- `/states/kube/state`
- `/states/db/state`

3 keys per role, 9 keys in total. 

If we split the storage into 3 secrets, i.e. `state`, `identity` and `replacement` each one of them will have the following format:

`state`:

```yaml
apiVersion: v1
data: |
            "kube": {
                "kind": "state",
                "version": "v2",
                ...
            }
           "app": {
                "kind": "state",
                "version": "v2",
                ...
            }
           "db": {
                "kind": "state",
                "version": "v2",
                ...
            }
    
kind: Secret
metadata:
  name: {.Release.Name}-state-{{$TELEPORT_REPLICA_NAME}}
  namespace: {.Release.Namespace}
```

`identity`:

```yaml
apiVersion: v1
data: |
            "kube": {
                "kind": "identity",
                "version": "v2",
                "metadata": {
                    "name": "current"
                },
                "spec": {
                    "key": "{key_content}",
                    "ssh_cert": "{ssh_cert_content}",
                    "tls_cert": "{tls_cert_content}",
                    "tls_ca_certs": ["{tls_ca_certs}"],
                    "ssh_ca_certs": ["{ssh_ca_certs}"]
                }
            }
            "db": {
                "kind": "identity",
                "version": "v2",
                "metadata": {
                    "name": "current"
                },
                "spec": {
                    "key": "{key_content}",
                    "ssh_cert": "{ssh_cert_content}",
                    "tls_cert": "{tls_cert_content}",
                    "tls_ca_certs": ["{tls_ca_certs}"],
                    "ssh_ca_certs": ["{ssh_ca_certs}"]
                }
            }
            "app": {
                "kind": "identity",
                "version": "v2",
                "metadata": {
                    "name": "current"
                },
                "spec": {
                    "key": "{key_content}",
                    "ssh_cert": "{ssh_cert_content}",
                    "tls_cert": "{tls_cert_content}",
                    "tls_ca_certs": ["{tls_ca_certs}"],
                    "ssh_ca_certs": ["{ssh_ca_certs}"]
                }
            }
    
kind: Secret
metadata:
  name: {.Release.Name}-identity-{{$TELEPORT_REPLICA_NAME}}
  namespace: {.Release.Namespace}
```

`replacement`:

```yaml
apiVersion: v1
data: |
            "kube": {
                "kind": "identity",
                "version": "v2",
                "metadata": {
                    "name": "current"
                },
                "spec": {
                    "key": "{key_content}",
                    "ssh_cert": "{ssh_cert_content}",
                    "tls_cert": "{tls_cert_content}",
                    "tls_ca_certs": ["{tls_ca_certs}"],
                    "ssh_ca_certs": ["{ssh_ca_certs}"]
                }
            }
            "db": {
                "kind": "identity",
                "version": "v2",
                "metadata": {
                    "name": "current"
                },
                "spec": {
                    "key": "{key_content}",
                    "ssh_cert": "{ssh_cert_content}",
                    "tls_cert": "{tls_cert_content}",
                    "tls_ca_certs": ["{tls_ca_certs}"],
                    "ssh_ca_certs": ["{ssh_ca_certs}"]
                }
            }
            "app": {
                "kind": "identity",
                "version": "v2",
                "metadata": {
                    "name": "current"
                },
                "spec": {
                    "key": "{key_content}",
                    "ssh_cert": "{ssh_cert_content}",
                    "tls_cert": "{tls_cert_content}",
                    "tls_ca_certs": ["{tls_ca_certs}"],
                    "ssh_ca_certs": ["{ssh_ca_certs}"]
                }
            }
    
kind: Secret
metadata:
  name: {.Release.Name}-replacement-{{$TELEPORT_REPLICA_NAME}}
  namespace: {.Release.Namespace}
```

The difference in the structure between the 3 secrets version and the single secret version is almost none and the second is simpler to handle than the first.

Reply:
Yeah I understand that, my concern is that we're basically trying to emulate backend structure within a Kubernetes secret which I'm not a fan of. In your example above btw, I also don't think that we should try and stuff all identities the same secret. I think it should be one identity per secret, maybe encoding role in the name or something. So in your example above you'll actually have 9 secrets, which I think is not a big deal.

To try and explain where I'm coming from advocating for this design - I think that _ideally_ (at least the way I imagined it when I was thinking about it) an identity would actually be its own custom resource e.g. `kind: TeleportIdentity`. Then it would be easy to reason about and you could in theory do `kubectl get identities` which is a standard way of extending K8s API. There's no real benefit to making identity into its own custom resource for now though, which is why a secret is the next best thing.

Another point in favor of keeping identities independent, one per object, is fewer potential issues if you only need to rotate, say, SSH CA or a database CA, as we wouldn't need to worry about concurrent updates to the same secret or something.

So I still think that we should just not really try to model it so it resembles existing backends.

Reply:
Got it.

I updated the RFD to incorporate those changes but I still have a lot of concerns about generating such a high number of secrets when secrets support natively multiple keys (`StringData map[string]string`).  
For instance, if we are running high availability with only 3 replicas it can generate up to 27 secrets, which is a lot from my perspective. This means that if the operator wants to upgrade the installation to switch the cluster to which the agent connects (different auth service) by providing a different invite token and proxy URL, he has to manually delete every secret which is painful since we generate such a high amount of secrets.

> To try and explain where I'm coming from advocating for this design - I think that ideally (at least the way I imagined it when I was thinking about it) an identity would actually be its own custom resource e.g. `kind: TeleportIdentity`. Then it would be easy to reason about and you could in theory do kubectl get identities which is a standard way of extending K8s API. There's no real benefit to making identity into its own custom resource for now though, which is why a secret is the next best thing.

I agree with you on `kind: TeleportIdentity` but, from my perspective, in that case, `current` and `replacement` identities must be stored together since they belong to the same agent role and we cannot dissociate them. From my perspective, they represent the same "identity" but in a different stage. We can argue about the fact that after rotation finishes  (successfully or not) we no longer require the `replacement` identity and they can be wiped out.

> Another point in favor of keeping identities independent, one per object, is fewer potential issues if you only need to rotate, say, SSH CA or a database CA, as we wouldn't need to worry about concurrent updates to the same secret or something.

This must also be present because if MachineID can rotate the credentials, then the version control must happen in order to prevent the MachineID bot and agent from racing on editing the secret.
For multiple roles running in the same agent, the behavior is the same as in the `SQLite` backend, where the concurrent conns to the DB are limited to 1 in order as described in:

https://github.com/gravitational/teleport/blob/bbcd832b75375ba3da05342c5b0c254189ac0ea0/lib/backend/lite/lite.go#L230-L235

This is easily achieved by using a mutex to protect the access which was what I initially thought.